indexer/cell_coverer.hpp

```diff
@@ -2,54 +2,58 @@
 
 #include "indexer/cell_id.hpp"
 
+#include "base/buffer_vector.hpp"
+
 #include "std/queue.hpp"
 #include "std/vector.hpp"
 
 // TODO: Move neccessary functions to geometry/covering_utils.hpp and delete this file.
 
+constexpr int SPLIT_RECT_CELLS_COUNT = 512;
+
 template <typename BoundsT, typename CellIdT>
-inline void SplitRectCell(CellIdT id,
-                          double minX, double minY,
-                          double maxX, double maxY,
-                          vector<CellIdT> & result)
+inline int SplitRectCell(CellIdT const & id, m2::RectD const & rect,
+                         array<pair<CellIdT, m2::RectD>, 4> & result)
 {
+  int index = 0;
   for (int8_t i = 0; i < 4; ++i)
   {
-    CellIdT child = id.Child(i);
+    CellIdT const child = id.Child(i);
     double minCellX, minCellY, maxCellX, maxCellY;
     CellIdConverter<BoundsT, CellIdT>::GetCellBounds(child, minCellX, minCellY, maxCellX, maxCellY);
-    if (!((maxX < minCellX) || (minX > maxCellX) || (maxY < minCellY) || (minY > maxCellY)))
-      result.push_back(child);
+
+    m2::RectD const childRect(minCellX, minCellY, maxCellX, maxCellY);
+    if (rect.IsIntersect(childRect))
+      result[index++] = {child, childRect};
   }
+  return index;
 }
 
 template <typename BoundsT, typename CellIdT>
-inline void CoverRect(double minX, double minY,
-                      double maxX, double maxY,
-                      size_t cells_count, int maxDepth,
-                      vector<CellIdT> & cells)
+inline void CoverRect(m2::RectD rect, size_t cellsCount, int maxDepth,
```

Comment:
Previously it wasn't assumed that cells is empty, now you ASSERT that result must be empty. Are you sure that all call places will be happy with that?